File: Semestrul 4/Medii de proiectare si programare/Laboratoare/FirmaTransportJava/Networking/src/main/java/transport/networking/protocol/TransportClientWorker.java
package transport.networking.protocol;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import transport.model.Seat;
import transport.model.Trip;
import transport.networking.dto.DTOUtils;
import transport.networking.dto.ReservationDTO;
import transport.networking.dto.SearchDTO;
import transport.networking.dto.TripDTO;
import transport.networking.dto.UserDTO;
import transport.services.IObserver;
import transport.services.ITransportService;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.net.Socket;
import java.util.List;

public class TransportClientWorker implements Runnable, IObserver {

    private final ITransportService server;
    private final Socket connection;

    private ObjectInputStream input;
    private ObjectOutputStream output;

    private volatile boolean connected;

    private static final Logger logger = LogManager.getLogger(TransportClientWorker.class);

    public TransportClientWorker(ITransportService server, Socket connection) {
        this.server = server;
        this.connection = connection;
        try {
            output = new ObjectOutputStream(connection.getOutputStream());
            output.flush();
            input = new ObjectInputStream(connection.getInputStream());
            connected = true;
        } catch (IOException e) {
            logger.error("Error initializing client worker", e);
        }
    }

    public void run() {
        while (connected) {
            try {
                Object request = input.readObject();
                Response response = handleRequest((Request) request);
                if (response != null) {
                    sendResponse(response);
                }
            } catch (Exception e) {
                logger.error("Error while handling request", e);
            }

            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                logger.error("Worker interrupted", e);
            }
        }

        try {
            input.close();
            output.close();
            connection.close();
        } catch (IOException e) {
            logger.error("Error closing client worker resources", e);
        }
    }

    private Response handleRequest(Request request) {
        try {
            switch (request.getType()) {
                case LOGIN:
                    server.login(DTOUtils.getFromDTO((UserDTO) request.getData()), this);
                    return new Response(ResponseType.OK, null);

                case LOGOUT:
                    server.logout(DTOUtils.getFromDTO((UserDTO) request.getData()));
                    connected = false;
                    return new Response(ResponseType.OK, null);

                case GET_TRIPS:
                    List<Trip> trips = server.getAllTrips();
                    return new Response(ResponseType.TRIPS_LIST, trips);

                case SEARCH_TRIPS:
                    List<Trip> filteredTrips = server.searchTrips(DTOUtils.getFromDTO((SearchDTO) request.getData()));
                    return new Response(ResponseType.SEARCH_RESULT, filteredTrips);

                case GET_SEATS_FOR_TRIP:
                    List<Seat> seats = server.getSeatsForTrip(DTOUtils.getFromDTO((TripDTO) request.getData()));
                    return new Response(ResponseType.SEATS_LIST, seats);

                case MAKE_RESERVATION:
                    server.makeReservation(DTOUtils.getFromDTO((ReservationDTO) request.getData()));
                    return new Response(ResponseType.OK, null);

                case CANCEL_RESERVATION:
                    server.cancelReservation(DTOUtils.getFromDTO((ReservationDTO) request.getData()));
                    return new Response(ResponseType.OK, null);

                default:
                    return new Response(ResponseType.ERROR, "Unknown request type");
            }
        } catch (Exception e) {
            return new Response(ResponseType.ERROR, e.getMessage());
        }
    }

    private void sendResponse(Response response) throws IOException {
        logger.debug("Sending response {}", response);
        synchronized (output) {
            output.writeObject(response);
            output.flush();
        }
    }

    @Override
    public void tripsUpdated(List<Trip> trips) {
        try {
            sendResponse(new Response(ResponseType.UPDATE, trips));
        } catch (IOException e) {
            logger.error("Error sending update", e);
        }
    }
}

File: Semestrul 4/Medii de proiectare si programare/Laboratoare/FirmaTransportC#/Network/protocol/TransportClientWorker.cs
using FirmaTransportC_.Model;
using FirmaTransportC_.Networking.dto;
using FirmaTransportC_.Services;
using log4net;
using System.Net.Sockets;
using System.Text;
using System.Text.Json;

namespace FirmaTransportC_.Networking.protocol
{
    public class TransportClientWorker : IObserver
    {
        private ITransportService server;
        private TcpClient connection;

        private NetworkStream stream;

        private volatile bool connected;

        private static readonly ILog log = LogManager.GetLogger("TransportClientWorker");

        public TransportClientWorker(ITransportService server, TcpClient connection)
        {
            this.server = server;
            this.connection = connection;
            try
            {
                stream = connection.GetStream();
                connected = true;
            }
            catch (Exception e)
            {
                log.Error("Error getting stream: " + e.Message);
            }
        }

        public virtual void run()
        {
            using StreamReader reader = new StreamReader(stream, Encoding.UTF8);
            while (connected)
            {
                try
                {
                    string requestJson = reader.ReadLine();

                    if (string.IsNullOrEmpty(requestJson))
                        continue;

                    log.DebugFormat("Received json request {0}", requestJson);

                    Request request = JsonSerializer.Deserialize<Request>(requestJson);

                    log.DebugFormat("Deserializaed Request {0} ", request);

                    Response response = handleRequest(request);

                    if (response != null)
                    {
                        sendResponse(response);
                    }
                }
                catch (Exception e)
                {
                    log.ErrorFormat("run eroare {0}", e.Message);
                    if (e.InnerException != null)
                        log.ErrorFormat("run inner error {0}", e.InnerException.Message);
                    log.Error(e.StackTrace);
                }

                try
                {
                    Thread.Sleep(1000);
                }
                catch (Exception e)
                {
                    log.Error(e.StackTrace);
                }
            }
            try
            {
                stream.Close();
                connection.Close();
            }
            catch (Exception e)
            {
                log.Error("Error " + e);
            }
        }

        private Response handleRequest(Request request)
        {
            try
            {
                switch (request.Type)
                {
                    case RequestType.LOGIN:
                        {
                            log.Debug("Login request received");
                            User user = DTOUtils.GetFromDTO(request.User);
                            lock (server)
                            {
                                server.Login(user, this);
                            }
                            return JsonProtocolUtils.CreateLoginResponse(user);
                        }
                    case RequestType.LOGOUT:
                        {
                            log.Debug("Logout request received");
                            User user = DTOUtils.GetFromDTO(request.User);
                            lock (server)
                            {
                                server.Logout(user);
                            }
                            connected = false;
                            return JsonProtocolUtils.CreateOkResponse();
                        }
                    case RequestType.GET_TRIPS:
                        {
                            log.Debug("Get trips request received");
                            List<Trip> trips;
                            lock (server)
                            {
                                trips = server.GetAllTrips();
                            }
                            return JsonProtocolUtils.CreateGetTripsResponse(trips);
                        }
                    case RequestType.SEARCH_TRIPS:
                        {
                            log.Debug("Search trips request received");
                            Trip search = DTOUtils.GetFromDTO(request.SearchParams);
                            List<Trip> trips;
                            lock (server)
                            {
                                trips = server.SearchTrips(search);
                            }

                            return JsonProtocolUtils.CreateSearchTripsResponse(trips);
                        }
                    case RequestType.GET_SEATS_FOR_TRIP:
                        {
                            log.Debug("Get seats for trip request received");
                            Trip trip = DTOUtils.GetFromDTO(request.Trip);
                            List<Seat> seats;
                            lock (server)
                            {
                                seats = server.GetSeatsForTrip(trip);
                            }

                            return JsonProtocolUtils.CreateGetSeatsResponse(seats);
                        }
                    case RequestType.MAKE_RESERVATION:
                        {
                            log.Debug("Make reservation request received");
                            Reservation reservation = DTOUtils.GetFromDTO(request.Reservation);
                            lock (server)
                            {
                                server.MakeReservation(reservation);
                            }
                            return JsonProtocolUtils.CreateOkResponse();
                        }
                    case RequestType.CANCEL_RESERVATION:
                        {
                            log.Debug("Cancel reservation request received");
                            Reservation reservation = DTOUtils.GetFromDTO(request.Reservation);
                            lock (server)
                            {
                                server.CancelReservation(reservation);
                            }
                            return new Response { Type = ResponseType.OK };
                        }
                }
                return new Response { Type = ResponseType.ERROR, ErrorMessage = "Unknown request type" };

            }
            catch (Exception e)
            {
                log.Error("Error handling request: " + e.Message);
                return new Response { Type = ResponseType.ERROR, ErrorMessage = e.Message };
            }
        }

        private void sendResponse(Response response)
        {
            try
            {
                string jsonString = JsonSerializer.Serialize(response);
                log.DebugFormat("Sending response {0}", jsonString);
                lock (stream)
                {
                    byte[] data = Encoding.UTF8.GetBytes(jsonString + "\n");
                    stream.Write(data, 0, data.Length);
                    stream.Flush();
                }
            }
            catch (Exception e)
            {
                log.Error("Error sending response: " + e.Message);
            }
        }

        public void tripsUpdated(List<Trip> trips)
        {
            try
            {
                sendResponse(JsonProtocolUtils.CreateTripsUpdatedResponse(trips));
            }
            catch (Exception e)
            {
                {
                    log.Error("Error sending trips updated response: " + e.Message);
                }
            }
        }
    }
}
